File: sonnet-java/app/src/main/java/org/example/token/Tokenizer.java
package org.example.token;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Tokenizer {
	private final Pattern idPattern = Pattern.compile("[a-zA-Z0-9!@$%^&/\\-_+=:<>.?*']+");

	public List<Token> tokenize(String input) {
		List<Token> tokens = new ArrayList<>();
		// tokens.add(new Token.OpeningParen());
		int length = input.length();
		int pos = 0;

		while (pos < length) {
			pos = skipWhitespaceAndComments(input, pos, length);
			if (pos >= length) break;

			char ch = input.charAt(pos);

			switch (ch) {
				case '(':
					tokens.add(new Token.OpeningParen());
					pos++;
					break;

				case ')':
					tokens.add(new Token.ClosingParen());
					pos++;
					break;

				case '"': {
					pos++;
					int start = pos;
					while (pos < length && input.charAt(pos) != '"') {
						pos++;
					}
					if (pos >= length) throw new TokenizerError();

					tokens.add(new Token.StringToken(input.substring(start, pos)));
					pos++;
					break;
				}

				case '0':
				case '1':
				case '2':
				case '3':
				case '4':
				case '5':
				case '6':
				case '7':
				case '8':
				case '9': {
					int start = pos;
					boolean hasDot = false;

					while (pos < length) {
						char curr = input.charAt(pos);
						if (curr == '.') {
							if (hasDot) throw new TokenizerError();
							hasDot = true;
						} else if (isSequencingOmen(curr)) {
							break;
						} else if (!Character.isDigit(curr)) {
							throw new TokenizerError();
						}
						pos++;
					}

					String lexeme = input.substring(start, pos);

					if (hasDot) {
						tokens.add(new Token.FloatToken(Double.parseDouble(lexeme)));
					} else {
						tokens.add(new Token.IntToken(Integer.parseInt(lexeme)));
					}

					break;
				}

				default: {
					int start = pos;
					while (pos < length) {
						Matcher m = idPattern.matcher(String.valueOf(input.charAt(pos)));
						if (!m.matches()) break;
						pos++;
					}

					String lexeme = input.substring(start, pos);
					if (lexeme.isEmpty()) throw new TokenizerError();
					tokens.add(new Token.IdentifierToken(lexeme));
					break;
				}
			}
		}
		// tokens.add(new Token.ClosingParen());
		return tokens;
	}

	private boolean isSequencingOmen(char c) {
		return Character.isWhitespace(c) || c == '(' || c == ')';
	}

	private int skipWhitespaceAndComments(String input, int pos, int length) {
		while (pos < length) {
			char ch = input.charAt(pos);
			if (Character.isWhitespace(ch)) {
				pos++;
			} else if (ch == '#') {
				while (pos < length && input.charAt(pos) != '\n') {
					pos++;
				}
			} else {
				break;
			}
		}
		return pos;
	}
}

File: sonnet-scala/src/token/Tokenizer.scala
package token

import app.CompilerError
import scala.collection.mutable.ListBuffer
import scala.util.matching.Regex

sealed trait Token

object Token {
    case class QuoteToken()                  extends Token
    case class IntToken(value: Int)          extends Token
    case class DoubleToken(value: Double)    extends Token
    case class StringToken(value: String)    extends Token
    case class IdentifierToken(name: String) extends Token
    case class OpeningParen()                extends Token
    case class ClosingParen()                extends Token
}

class TokenizerError extends CompilerError("Tokenizer", null)

val idPattern: Regex = """[a-zA-Z0-9!@$%^&/\-_+=:<>.?*']+""".r

def tokenize(input: String): List[Token] = {
    val tokens = ListBuffer[Token]()
    val length = input.length
    var pos    = 0

    while (pos < length) {
        pos = skipWhitespaceAndComments(input, pos, length)
        if (pos < length) {
            val ch = input.charAt(pos)

            ch match {
                case '(' => {
                    tokens += Token.OpeningParen()
                    pos += 1
                }

                case ')' => {
                    tokens += Token.ClosingParen()
                    pos += 1
                }

                case '\'' => {
                    tokens += Token.QuoteToken()
                    pos += 1
                }

                case '"' => {
                    pos += 1
                    val start = pos
                    while (pos < length && input.charAt(pos) != '"') {
                        pos += 1
                    }

                    if (pos >= length) {
                        throw TokenizerError()
                    }

                    tokens += Token.StringToken(input.substring(start, pos))
                    pos += 1
                }

                case '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' => {
                    val start     = pos
                    var hasDot    = false
                    var breakLoop = false

                    while (pos < length && !breakLoop) {
                        val curr = input.charAt(pos)
                        if (curr == '.') {
                            if (hasDot) {
                                throw TokenizerError()
                            }
                            hasDot = true
                            pos += 1
                        } else if (isSequencingOmen(curr)) {
                            breakLoop = true
                        } else if (curr.isDigit) {
                            pos += 1
                        } else {
                            throw TokenizerError()
                        }
                    }

                    val lexeme = input.substring(start, pos)
                    if (hasDot) {
                        tokens += Token.DoubleToken(lexeme.toDouble)
                    } else {
                        tokens += Token.IntToken(lexeme.toInt)
                    }
                }

                case _ => {
                    val start     = pos
                    var breakLoop = false

                    while (pos < length && !breakLoop) {
                        val currCharStr = input.charAt(pos).toString
                        if (idPattern.matches(currCharStr)) {
                            pos += 1
                        } else {
                            breakLoop = true
                        }
                    }

                    val lexeme = input.substring(start, pos)
                    if (lexeme.isEmpty) {
                        throw TokenizerError()
                    }
                    tokens += Token.IdentifierToken(lexeme)
                }
            }
        }
    }
    tokens.toList
}

def isSequencingOmen(c: Char): Boolean = {
    c.isWhitespace || c == '(' || c == ')'
}

def skipWhitespaceAndComments(input: String, startPos: Int, length: Int): Int = {
    var pos       = startPos
    var breakLoop = false

    while (pos < length && !breakLoop) {
        val ch = input.charAt(pos)
        if (ch.isWhitespace) {
            pos += 1
        } else if (ch == '#') {
            while (pos < length && input.charAt(pos) != '\n') {
                pos += 1
            }
        } else {
            breakLoop = true
        }
    }
    pos
}
